### PR TITLE
Make `Clean` take &mut DocContext

### PR DESCRIPTION
**Pull request by jyn514**
- Take `FnMut` in `rustc_trait_selection::find_auto_trait_generics`
- Take `&mut DocContext` in most of `clean`
- Collect the iterator in auto_trait_impls instead of iterating lazily; the lifetimes were really bad.

This combined with https://github.com/rust-lang/rust/pull/82018 should hopefully help with https://github.com/rust-lang/rust/pull/82014 by allowing `cx.cache.exported_traits` to be modified in `register_res`. Previously it had to use interior mutability, which required either adding a RefCell to `cache.exported_traits` on *top* of the existing `RefCell<Cache>` or mixing reads and writes between `cx.exported_traits` and `cx.cache.exported_traits`. I don't currently have that working but I expect it to be reasonably easy to add after this.

### CODE REVIEW

**Comment by jyn514** [comment]
Hmm, maybe r? @camelid? The changes are all fairly simple, there's just a lot of them :sweat_smile: sorry for all the work

File: src/librustdoc/clean/utils.rs [hunk 1]
```diff
@@ -421,29 +421,26 @@ crate fn resolve_type(cx: &DocContext<'_>, path: Path, id: hir::HirId) -> Type {
...
 crate fn get_auto_trait_and_blanket_impls(
-    cx: &DocContext<'tcx>,
+    cx: &mut DocContext<'tcx>,
     ty: Ty<'tcx>,
     param_env_def_id: DefId,
 ) -> impl Iterator<Item = Item> {
-    let auto_impls = cx
-        .sess()
-        .prof
+    let prof = &cx.tcx.sess.prof;
```
Comment (camelid) [inline src/librustdoc/clean/utils.rs:433]: What is the difference between these? Is it so the borrow-checker can see we're only using `.tcx.sess.prof` rather than all of `cx`?

**Review by camelid — COMMENTED** [review]
Left an initial review on some parts (haven't yet reviewed some of the files with massive changes like `clean/mod.rs`).

There were several places with statements like `let cx = self.cx;` or `let tcx = self.cx.tcx` and changes from `cx.sess()` to `cx.tcx.sess`. I'm assuming these changes are necessary to please the borrow-checker?

File: src/librustdoc/clean/auto_trait.rs [hunk 2]
```diff
@@ -21,42 +21,41 @@ struct RegionDeps<'tcx> {
...
     // FIXME(eddyb) figure out a better way to pass information about
     // parametrization of `ty` than `param_env_def_id`.
     crate fn get_auto_trait_impls(&self, ty: Ty<'tcx>, param_env_def_id: DefId) -> Vec<Item> {
-        let param_env = self.cx.tcx.param_env(param_env_def_id);
+        let tcx = self.tcx;
+        let param_env = tcx.param_env(param_env_def_id);
 
         debug!("get_auto_trait_impls({:?})", ty);
-        let auto_traits = self.cx.auto_traits.iter().cloned();
+        let auto_traits: Vec<_> = self.cx.borrow().auto_traits.iter().cloned().collect();
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:43]: Yes, otherwise `self.cx` is used both here and inside the `filter_map` closure at the same time.

File: src/librustdoc/clean/auto_trait.rs [hunk 3]
```diff
@@ -154,13 +148,6 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
             .clone()
     }
 
-    fn region_name(&self, region: Region<'_>) -> Option<Symbol> {
-        match region {
-            &ty::ReEarlyBound(r) => Some(r.name),
-            _ => None,
-        }
-    }
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:162]: Good catch, thanks. I'll change `get_lifetime` to be a free function.

File: src/librustdoc/clean/utils.rs (hunk 1 above)
Comment (jyn514) [reply src/librustdoc/clean/utils.rs:433]: Yes, exactly. `cx.sess()` takes a mutable borrow because the `&self` in the function doesn't cause the reference to decay to a &T soon enough I think (??) it gave me errors as if it thought all of `&mut self` was borrowed and not just the immutable reference. Let me find the original errors, one sec.

**Comment by jyn514** [comment]
> There were several places with statements like let cx = self.cx; or let tcx = self.cx.tcx and changes from cx.sess() to cx.tcx.sess. I'm assuming these changes are necessary to please the borrow-checker?

Yes.
```
error[E0500]: closure requires unique access to `self` but it is already borrowed
  --> src/librustdoc/clean/auto_trait.rs:43:25
   |
41 |         let auto_traits = self.cx.auto_traits.iter().cloned();
   |                           ------------------- borrow occurs here
42 |         auto_traits
43 |             .filter_map(|trait_def_id| {
   |              ---------- ^^^^^^^^^^^^^^ closure construction occurs here
   |              |
   |              first borrow later used by call
...
46 |                     substs: self.cx.tcx.mk_substs_trait(ty, &[]),
   |                             ---- second borrow occurs due to use of `self` in closure
```

File: src/librustdoc/clean/auto_trait.rs [hunk 4]
```diff
@@ -21,42 +21,41 @@ struct RegionDeps<'tcx> {
 }
 
 crate struct AutoTraitFinder<'a, 'tcx> {
-    crate cx: &'a core::DocContext<'tcx>,
+    crate cx: RefCell<&'a mut core::DocContext<'tcx>>,
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:24]: > Why is this use of RefCell being added?

Ugh, this whole bit was a mess. I couldn't figure out how to get it to work, this is the error without the RefCell:

```
error[E0500]: closure requires unique access to `self` but it is already borrowed
  --> src/librustdoc/clean/auto_trait.rs:56:82
   |
56 |                     self.f.find_auto_trait_generics(ty, param_env, trait_def_id, |infcx, info| {
   |                     ------ ------------------------                              ^^^^^^^^^^^^^ closure construction occurs here
   |                     |      |
   |                     |      first borrow later used by call
   |                     borrow occurs here
...
70 |                         let new_generics = self.param_env_to_generics(
   |                                            ---- second borrow occurs due to use of `self` in closure

```

> Also, if a cell is needed, can't we use Cell instead since references are Copy, or are references 'special'?

&T is Copy, but &mut T is not.

File: src/librustdoc/clean/auto_trait.rs [hunk 5]
```diff
@@ -438,7 +424,7 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
     // * We explicitly add a '?Sized' bound if we didn't find any 'Sized' predicates for a type
     fn param_env_to_generics(
         &self,
-        tcx: TyCtxt<'tcx>,
+        cx: &mut DocContext<'tcx>,
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:427]: Yes: https://github.com/rust-lang/rust/pull/82020#discussion_r575917209

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:427]: Here's the error if I take `&self` *and* `&mut DocContext`:

```
error[E0502]: cannot borrow `self.cx` as mutable because it is also borrowed as immutable
  --> src/librustdoc/clean/auto_trait.rs:71:29
   |
70 |                         let new_generics = self.param_env_to_generics(
   |                                            ---- --------------------- immutable borrow later used by call
   |                                            |
   |                                            immutable borrow occurs here
71 |                             &mut self.cx,
   |                             ^^^^^^^^^^^^ mutable borrow occurs here

error[E0500]: closure requires unique access to `self` but it is already borrowed
  --> src/librustdoc/clean/auto_trait.rs:56:82
   |
56 |                     self.f.find_auto_trait_generics(ty, param_env, trait_def_id, |infcx, info| {
   |                     ------ ------------------------                              ^^^^^^^^^^^^^ closure construction occurs here
   |                     |      |
   |                     |      first borrow later used by call
   |                     borrow occurs here
...
70 |                         let new_generics = self.param_env_to_generics(
   |                                            ---- second borrow occurs due to use of `self` in closure
```

File: src/librustdoc/clean/auto_trait.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:24]: OMG I'm a dumbass, I just needed to make `f` a separate field from the struct. Give me a second.

File: src/librustdoc/clean/auto_trait.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs]: This has since been removed now that I figured out I was being dumb :)

File: src/librustdoc/clean/auto_trait.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs]: Ok this no longer uses RefCell at all :tada:

File: src/librustdoc/clean/utils.rs (hunk 1 above)
Comment (jyn514) [reply src/librustdoc/clean/utils.rs:433]: Ok my explanation above is wrong - the issue is `AutoTraitFInder::new` takes `&mut DocContext` but using `.sess()` ties the lifetime of the profiler to `DocContext` even though it shouldn't. Using `cx.tcx.sess` copies `tcx` instead and gives the session a lifetime of `'tcx`.

**Comment by jyn514** [comment]
Well, these were certainly necessary (https://github.com/rust-lang/rust/pull/82020#discussion_r575920250):

> statements like let tcx = self.cx.tcx and changes from cx.sess() to cx.tcx.sess. 

but these may have been extraneous:

> let cx = self.cx;

I'll go through the changes again and see if I can revert them.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 6]
```diff
@@ -434,9 +425,8 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
         path_str: &'a str,
         module_id: DefId,
     ) -> Result<Res, ResolutionFailure<'a>> {
-        let cx = self.cx;
         let path = ast::Path::from_ident(Ident::from_str(path_str));
-        cx.enter_resolver(|resolver| {
+        self.cx.enter_resolver(|resolver| {
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:429]: This change isn't strictly necessary, but it avoids having to write `let cx = &self.cx;`, which looks very silly since cx is a mutable reference.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 7]
```diff
@@ -948,12 +938,18 @@ impl LinkCollector<'_, '_> {
             return None;
         }
 
-        let cx = self.cx;
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:951]: This one is necessary.
```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
    --> src/librustdoc/passes/collect_intra_doc_links.rs:1082:39
     |
942  |         let cx = &self.cx;
     |                  -------- immutable borrow occurs here
...
1082 |         let (mut res, mut fragment) = self.resolve_with_disambiguator_cached(
     |                                       ^^^^ mutable borrow occurs here
...
1106 |                             cx,
     |                             -- immutable borrow later used here
```

File: src/librustdoc/passes/doc_test_lints.rs [hunk 8]
```diff
@@ -19,27 +19,20 @@ crate const CHECK_PRIVATE_ITEMS_DOC_TESTS: Pass = Pass {
...
     coll.fold_crate(krate)
 }
 
 impl<'a, 'tcx> DocFolder for PrivateItemDocTestLinter<'a, 'tcx> {
     fn fold_item(&mut self, item: Item) -> Option<Item> {
-        let cx = self.cx;
         let dox = item.attrs.collapsed_doc_value().unwrap_or_else(String::new);
 
-        look_for_tests(&cx, &dox, &item);
+        look_for_tests(self.cx, &dox, &item);
```
Comment (jyn514) [inline src/librustdoc/passes/doc_test_lints.rs:35]: Same as before, I could have done `let cx = &self.cx;` but it seemed silly when it's only used once.

File: src/librustdoc/clean/auto_trait.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:162]: FYI, you also had at least one other function that you changed from methods to associated functions.

File: src/librustdoc/clean/utils.rs (hunk 1 above)
Comment (camelid) [reply src/librustdoc/clean/utils.rs:433]: A bit over my head, but seems harmless :)

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 6 above)
Comment (camelid) [reply src/librustdoc/passes/collect_intra_doc_links.rs:429]: > This change isn't strictly necessary, but it avoids having to write `let cx = &self.cx;`, which looks very silly since cx is a mutable reference.

Wait... how can it be a mutable reference? Do you mean an `&&mut ...`? I thought you couldn't mutate through those...

File: src/librustdoc/passes/doc_test_lints.rs (hunk 8 above)
Comment (camelid) [reply src/librustdoc/passes/doc_test_lints.rs:35]: Yeah, I'm fine with getting rid of things like that since I think it makes the code better—although it does make reviewing a bit harder.

File: src/librustdoc/clean/auto_trait.rs [hunk 9]
```diff
@@ -471,9 +459,7 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
         let mut ty_to_fn: FxHashMap<Type, (Option<PolyTrait>, Option<Type>)> = Default::default();
 
         for p in clean_where_predicates {
-            let replaced = p.fold_with(&mut replacer);
-            let (orig_p, p) = (replaced, replaced.clean(cx));
-
+            let (orig_p, p) = (p, p.clean(self.cx));
```
Comment (camelid) [inline src/librustdoc/clean/auto_trait.rs]: This seems like a behavior change.

File: src/librustdoc/clean/auto_trait.rs [hunk 10]
```diff
@@ -423,8 +408,8 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
     // K', we use the dedicated syntax 'T: Fn() -> K'
     // * We explicitly add a '?Sized' bound if we didn't find any 'Sized' predicates for a type
     fn param_env_to_generics(
-        &self,
-        cx: &mut DocContext<'tcx>,
+        &mut self,
+        tcx: TyCtxt<'tcx>,
```
Comment (camelid) [inline src/librustdoc/clean/auto_trait.rs:412]: Why does this take `tcx`? It seems like it uses `self.cx.tcx` in the body.

Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:412]: Is it just because it took `tcx` before?

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 6 above)
Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:429]: You can't, everything that uses it only needs &DocContext (just like it did before). Essentially it's reborrowing the &&mut T as &&T at every use and then using the fact that & is Copy to get a &T.

**Comment by camelid** [comment]
Am I right in thinking that this will make getting rid of the `RefCell`s in `DocContext` a lot easier?

File: src/librustdoc/clean/auto_trait.rs (hunk 10 above)
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:412]: > Is it just because it took tcx before?

Yes, I was trying to keep the churn to a minimum.

**Comment by jyn514** [comment]
> Am I right in thinking that this will make getting rid of the `RefCell`s in `DocContext` a lot easier?

Much so, yes. That would be a good follow up pr, but I didn't want to do it here because it's already +-300.

**Comment by camelid** [comment]
> Much so, yes. That would be a good follow up pr, but I didn't want to do it here because it's already +-300.

Yeah, I might try doing it after this PR if that's okay with you :)

File: src/librustdoc/clean/auto_trait.rs (hunk 10 above)
Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:412]: That's fine, but I feel a bit confused—why is CI passing if the argument is unused? Or is it used somewhat in the body?

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 6 above)
Comment (camelid) [reply src/librustdoc/passes/collect_intra_doc_links.rs:429]: Oh... I see. 👍

File: src/librustdoc/clean/auto_trait.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:162]: That function is `handle_lifetimes`.

File: src/librustdoc/clean/auto_trait.rs (hunk 10 above)
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:412]: Yes, it's used even though you can get it from cx.tcx.

Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:412]: Okay, we can remove that later then.

Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:412]: Opened #82137.

File: src/librustdoc/clean/auto_trait.rs (hunk 9 above)
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs]: It's not actually - it's just that the predicate is folded above (in `clean_where_predicates`) instead of here. It looks confusing because you're only looking at 2 commits, not the full diff.

File: src/librustdoc/clean/auto_trait.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:162]: Changed it to a free function, thanks. I recommend reviewing that commit with `git show -w --reverse --color-moved=dimmed-zebra --color-moved-ws=allow-indentation-change`.

**Comment by camelid** [comment]
What do you think of undoing the move to free functions? It's probably better overall to use free functions, but it makes reviewing the diffs a lot harder. I used the `git show -w --reverse --color-moved=dimmed-zebra --color-moved-ws=allow-indentation-change` command, but it means I can't review easily on GitHub. And then you can always open a PR after this one to move them.

**Comment by jyn514** [comment]
@camelid I tried switching this to have all associated functions and got a lifetime error I don't understand:

```
error: lifetime may not live long enough
   --> src/librustdoc/clean/auto_trait.rs:144:9
    |
27  | impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
    |      --  ---- lifetime `'tcx` defined here
    |      |
    |      lifetime `'a` defined here
...
144 |         Self::region_name(region)
    |         ^^^^^^^^^^^^^^^^^ requires that `'tcx` must outlive `'a`
    |
    = help: consider adding the following bound: `'tcx: 'a`

error: lifetime may not live long enough
   --> src/librustdoc/clean/auto_trait.rs:202:24
    |
27  | impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
    |      --  ---- lifetime `'tcx` defined here
    |      |
    |      lifetime `'a` defined here
...
202 |                     if Self::region_name(r1) != Self::region_name(r2) {
    |                        ^^^^^^^^^^^^^^^^^ requires that `'tcx` must outlive `'a`
    |
    = help: consider adding the following bound: `'tcx: 'a`

error: aborting due to 2 previous errors
```

Would you be ok with keeping `region_name` as a free function even though it's inconsistent?

**Comment by camelid** [comment]
Yikes, that is an intimidating error!

> Would you be ok with keeping `region_name` as a free function and leaving the others as associated functions, even though it's inconsistent? i.e. just drop the last commit.

Sure, sounds fine :)

**Comment by camelid** [comment]
> > Would you be ok with keeping `region_name` as a free function and leaving the others as associated functions, even though it's inconsistent? i.e. just drop the last commit.
> 
> Sure, sounds fine :)

Even aside from the lifetimes issues, `region_name` is so small and general-purpose compared to at least some of the others that it's significantly better as a free function.

**Comment by camelid** [comment]
(FYI, you have very minor merge conflicts.)

**Comment by camelid** [comment]
I beat bors to it! 😆

**Comment by jyn514** [comment]
> changes from cx.sess() to cx.tcx.sess.

I reverted these now that `sess()` now properly returns `&'tcx Session` instead of `&'cx Session`.

**Comment by camelid** [comment]
Should we get a perf run on this after a little more reviewing?

File: src/librustdoc/clean/auto_trait.rs [hunk 11]
```diff
@@ -21,42 +21,38 @@ struct RegionDeps<'tcx> {
...
     }
 
     // FIXME(eddyb) figure out a better way to pass information about
     // parametrization of `ty` than `param_env_def_id`.
-    crate fn get_auto_trait_impls(&self, ty: Ty<'tcx>, param_env_def_id: DefId) -> Vec<Item> {
-        let param_env = self.cx.tcx.param_env(param_env_def_id);
+    crate fn get_auto_trait_impls(&mut self, ty: Ty<'tcx>, param_env_def_id: DefId) -> Vec<Item> {
+        let tcx = self.cx.tcx;
+        let param_env = tcx.param_env(param_env_def_id);
+        let f = auto_trait::AutoTraitFinder::new(self.cx.tcx);
```
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:37]: I can change it here if you like - the change is only needed inside the closure, because otherwise it will borrow all of `self`. See also https://github.com/rust-lang/rfcs/pull/2229.

**Comment by jyn514** [comment]
> Should we get a perf run on this after a little more reviewing?

I don't expect the perf to be much different, but it doesn't hurt to check.

@bors try @rust-timer queue

**Comment by jyn514** [comment]
Perf looks like noise.

**Comment by camelid** [comment]
> I don't expect the perf to be much different, but it doesn't hurt to check. Most of the perf improvements will come in follow-up PRs that remove RefCell.

> Perf looks like noise.

Yeah, I just wanted to make sure this didn't cause any regressions (like the change to use `.collect()` in the auto trait collector).

File: src/librustdoc/clean/auto_trait.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (camelid) [reply src/librustdoc/clean/auto_trait.rs:37]: Ah, I know about RFC 2229, but I forgot that it's still being implemented — maybe seeing the PRs that implement parts of it makes me think it's already finished 😆

**Review by camelid — COMMENTED** [review]
Heh, I thought reviewing `clean/mod.rs` would be the hardest but it was actually fairly straightforward — just a lot of `&DocContext` to `&mut DocContext`!

I think this is close, but since it's a fairly large change that reaches across the codebase, I would like someone else to take a look too. @GuillaumeGomez do you think could review this as well?

File: src/librustdoc/clean/auto_trait.rs [hunk 12]
```diff
@@ -21,42 +21,38 @@ struct RegionDeps<'tcx> {
...
 
         debug!("get_auto_trait_impls({:?})", ty);
-        let auto_traits = self.cx.auto_traits.iter().cloned();
+        let auto_traits: Vec<_> = self.cx.auto_traits.iter().cloned().collect();
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/auto_trait.rs:40]: Why did you change this? It's a bit surprising considering that you're using `into_iter` just after on `auto_traits`.

File: src/librustdoc/clean/auto_trait.rs [hunk 13]
```diff
@@ -437,7 +419,7 @@ impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
     // K', we use the dedicated syntax 'T: Fn() -> K'
     // * We explicitly add a '?Sized' bound if we didn't find any 'Sized' predicates for a type
     fn param_env_to_generics(
-        &self,
+        &mut self,
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/auto_trait.rs:422]: What is requiring it to be `mut` here? Can't find out just from reading the changes...

File: src/librustdoc/clean/blanket_impl.rs [hunk 14]
```diff
@@ -10,17 +10,13 @@ use rustc_span::DUMMY_SP;
...
 
 impl<'a, 'tcx> BlanketImplFinder<'a, 'tcx> {
-    crate fn new(cx: &'a core::DocContext<'tcx>) -> Self {
-        BlanketImplFinder { cx }
-    }
-
     // FIXME(eddyb) figure out a better way to pass information about
     // parametrization of `ty` than `param_env_def_id`.
-    crate fn get_blanket_impls(&self, ty: Ty<'tcx>, param_env_def_id: DefId) -> Vec<Item> {
+    crate fn get_blanket_impls(&mut self, ty: Ty<'tcx>, param_env_def_id: DefId) -> Vec<Item> {
```
Comment (GuillaumeGomez) [inline src/librustdoc/clean/blanket_impl.rs:19]: Why this change?

File: src/librustdoc/clean/auto_trait.rs (hunk 12 above)
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:40]: This borrows from `cx.auto_traits`, which prevents `cx` from being used in the closure because `.clean(cx)` (on line 106) needs unique access.

File: src/librustdoc/clean/auto_trait.rs (hunk 13 above)
Comment (jyn514) [reply src/librustdoc/clean/auto_trait.rs:422]: `p.clean(self.cx)` on line 473 takes `&mut DocContext`.

File: src/librustdoc/clean/blanket_impl.rs (hunk 14 above)
Comment (jyn514) [reply src/librustdoc/clean/blanket_impl.rs:19]: `clean(self.cx)` (on line 117) takes `&mut DocContext`.

**Comment by GuillaumeGomez** [comment]
All good for me as well, thanks!

@bors: r=camelid,GuillaumeGomez

**Comment by camelid** [comment]
Will hopefully start on removing `RefCell`s over the next few days! 🎉

**Comment by jyn514** [comment]
@bors retry

Reported that this keeps timing out, I find it hard to believe it could be related to this change: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/i686-gnu*.20jobs.20keep.20timing.20out

**Comment by camelid** [comment]
> A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1938000148) [(plain)](https://github.com/rust-lang-ci/rust/commit/9b471a3f5fe57e5c6e08acf665f2094422415a3d/checks/1938000148/logs)
> Click to see the possible cause of the failure (guessed by this bot)
> 
> ```
>   git switch -
> 
> Turn off this advice by setting config variable advice.detachedHead to false
> 
> HEAD is now at 9b471a3f5 Auto merge of #82020 - jyn514:mut-passes, r=camelid,GuillaumeGomez
> ##[group]Run src/ci/publish_toolstate.sh
> src/ci/publish_toolstate.sh
> env:
>   SCCACHE_BUCKET: rust-lang-ci-sccache2
>   DEPLOY_BUCKET: rust-lang-ci2
>   TOOLSTATE_REPO: https://github.com/rust-lang-nursery/rust-toolstate
> ---
>   CACHE_DOMAIN: ci-caches.rust-lang.org
>   TOOLSTATE_REPO_ACCESS_TOKEN: ***
> ##[endgroup]
> Cloning into 'rust-toolstate'...
> /home/runner/work/rust/rust/src/tools/publish_toolstate.py:121: DeprecationWarning: 'U' mode is deprecated
> 📣 Toolstate changed by rust-lang/rust#82020!
>   with open(path, 'rU') as f:
> 
> Tested on commit rust-lang/rust@9b471a3f5fe57e5c6e08acf665f2094422415a3d.
> Direct link to PR: <https://github.com/rust-lang/rust/pull/82020>
> 
> 💔 miri on windows: test-fail → build-fail (cc @RalfJung @eddyb @oli-obk).
> 💔 miri on linux: test-fail → build-fail (cc @RalfJung @eddyb @oli-obk).
> Traceback (most recent call last):
> Traceback (most recent call last):
>   File "/home/runner/work/rust/rust/src/tools/publish_toolstate.py", line 338, in <module>
>     response = urllib2.urlopen(urllib2.Request(
>   File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib/python3.8/urllib/request.py", line 522, in open
>     req = meth(req)
>   File "/usr/lib/python3.8/urllib/request.py", line 1281, in do_request_
>     raise TypeError(msg)
> TypeError: POST data should be bytes, an iterable of bytes, or a file object. It cannot be of type str.
> ##[error]Process completed with exit code 1.
> ```

Seems like a bug in the toolstate program. cc @rust-lang/infra

**Comment by camelid** [comment]
Also, it seems surprising that Miri no longer builds after this PR. The only rustc change shouldn't affect Miri and should be backwards-compatible.

**Comment by jyn514** [comment]
@camelid this is a known issue:https://github.com/rust-lang/rust/issues/82254
